### PR TITLE
[docs] feat: Add NeMo-RL & verl E2E testing skill

### PR DESCRIPTION
**Pull request by HollowMan6**
# What does this PR do ?

Adds new agent skills for validating Megatron-Bridge model/provider changes through both NeMo-RL's Megatron backend and verl's Megatron backend.

# Changelog

- Add `skills/nemo-rl-e2e-testing/SKILL.md` with the external NeMo-RL E2E validation workflow.
- Add `skills/verl-e2e-testing/SKILL.md` with the external verl E2E validation workflow.
- Document NeMo-RL Level 0-6 coverage guidance for Megatron policy GRPO smoke, LoRA/checkpoint variants, non-colocated vLLM refit, optional Megatron
generation, parallelism stress, architecture-specific checks, and learning-signal runs.
- Document verl Level 0-5 coverage guidance for LoRA + DDP smoke, save/resume, parallelism stress, Megatron-FSDP, architecture-specific checks, and
learning-signal runs.
- Add setup guidance for Bridge/downstream checkout paths, `PYTHONPATH`, data/model/cache paths, reward handling, and Slurm/container runs.
- Add failure triage and pass criteria for common downstream integration issues such as missing runtime dependencies, stale attach containers, storage
limits, and post-completion shutdown noise.
- Link both new skills from `docs/skills-index.md`.
- Reference both E2E workflows from `skills/adding-model-support/tests-and-examples.md`.

# GitHub Actions CI

See the [CI section](https://github.com/NVIDIA-NeMo/Megatron-Bridge/blob/main/CONTRIBUTING.md#-running-github-ci)in the Contributing doc for how to trigger the CI. A Nvidia developer will need to approve and trigger the CI for external contributors.

# Before your PR is "Ready for review"

**Pre checks**:

- [x] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA-NeMo/Megatron-Bridge/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [x] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?

If you haven't finished some of the above items you can still open "Draft" PR.

# Additional Information

- Related to # (issue)

image

### CODE REVIEW

File: skills/verl-e2e-testing/SKILL.md [hunk 1]
```diff
@@ -0,0 +1,438 @@
...
+```
+
+Capture logs to a file for review:
+
+```bash
+mkdir -p "${LOG_DIR:-$PWD/verl_e2e_logs}"
+LOG_FILE="${LOG_DIR:-$PWD/verl_e2e_logs}/verl_lora_ddp_$(date +%Y%m%d_%H%M%S).log"
+<same-env-as-above> bash tests/special_e2e/run_ppo_trainer_megatron.sh 2>&1 | tee "$LOG_FILE"
+grep -E "Training Progress|VANILLA_MBRIDGE|Traceback|RuntimeError|KeyError|ValueError" "$LOG_FILE"
+```
```
[the remark this answers is not in the document]
Comment (HollowMan6) [reply skills/verl-e2e-testing/SKILL.md:263]: fixed

File: docs/skills-index.md [hunk 2]
```diff
@@ -26,6 +26,8 @@ skills/adding-model-support/llm-patterns
 skills/adding-model-support/vlm-patterns
 skills/adding-model-support/recipe-patterns
 skills/adding-model-support/tests-and-examples
+skills/verl-e2e-testing/SKILL
+skills/nemo-rl-e2e-testing/SKILL
 ```
```
[the remark this answers is not in the document]
Comment (HollowMan6) [reply docs/skills-index.md:31]: Updated the PR description.

File: skills/nemo-rl-e2e-testing/SKILL.md [hunk 3]
```diff
@@ -0,0 +1,523 @@
...
+This is not a replacement for Bridge model parity tests. A NeMo-RL GRPO or SFT run proves that Bridge can survive an external RL training loop; architecture correctness still comes from Bridge import/export, logits, roundtrip, and model-specific inference tests.
+
+## Scope
+
+Think in coverage levels. Start with Level 0 and add only the levels justified by the change.
+
+| Level | Required when | What it proves |
+|---|---|---|
+| 0: Megatron policy GRPO smoke | Any new provider or provider config change that claims NeMo-RL compatibility | NeMo-RL can import the local Bridge provider, build a Megatron policy, initialize optimizer/scheduler state, run rollout/ref/logprob wiring, and finish a short GRPO job |
+| 1: PEFT + checkpoint | PEFT, adapter loading/saving, checkpointing, HF export, optimizer state, or resume behavior changed | NeMo-RL can apply Bridge PEFT hooks, save through NeMo-RL checkpoint scheduling, and resume without losing training state |
```
Comment (yaoyu-33) [inline skills/nemo-rl-e2e-testing/SKILL.md]: not sure if nemo rl peft is always working?

File: skills/nemo-rl-e2e-testing/SKILL.md [hunk 4]
```diff
@@ -0,0 +1,523 @@
...
+| 2: Non-colocated vLLM refit | HF export, weight mapping, policy-to-generation refit, delta compression, packed transfer, or vLLM update behavior changed | Bridge-exported weights can be transferred from the Megatron policy worker into separate vLLM generation workers |
+| 3: Megatron generation backend | Bridge generation model construction, mcore generation config, KV/cache behavior, or Megatron inference path changed | NeMo-RL can use Megatron for both policy and generation rather than only vLLM generation |
```
Comment (yaoyu-33) [inline skills/nemo-rl-e2e-testing/SKILL.md]: not sure if megatron inference is relevant

File: skills/verl-e2e-testing/SKILL.md [hunk 5]
```diff
@@ -0,0 +1,437 @@
...
+## Scope
+
+Think in coverage levels. Start with Level 0 and add only the levels justified by the change.
+
+| Level | Required when | What it proves |
+|---|---|---|
+| 0: LoRA + DDP smoke | Any new provider or provider config change that claims verl compatibility | verl can import the local Bridge provider, apply PEFT, wrap with Megatron DDP, build optimizer state, run rollout/ref/critic wiring, and finish one PPO step |
+| 1: Save/resume | PEFT, checkpointing, HF export, adapter export, optimizer state, or resume behavior changed | verl-owned checkpoint scheduling can save and reload Bridge-built model state |
+| 2: Parallelism stress | Provider finalization, mpu-derived settings, TP/PP/CP/EP, sequence parallel, or dispatcher behavior changed | provider settings remain correct under non-trivial Megatron parallel state |
+| 3: Megatron-FSDP | FSDP wrapping, sharding, checkpoint format, or distributed optimizer behavior changed | the same provider works when verl selects Megatron-FSDP instead of DDP |
```
Comment (yaoyu-33) [inline skills/verl-e2e-testing/SKILL.md]: fsdp not needed today

File: skills/nemo-rl-e2e-testing/SKILL.md (hunk 3 above)
Comment (HollowMan6) [reply skills/nemo-rl-e2e-testing/SKILL.md]: Now I made this conditional instead of assuming NeMo-RL PEFT always works: Level 1 is now “LoRA/checkpoint variant” and only requires PEFT when the NeMo-RL checkout supports that path. I also added guidance to mark PEFT as not applicable/blocked by NeMo-RL when unsupported, while still using Level 0 GRPO as the provider compatibility smoke.

File: skills/nemo-rl-e2e-testing/SKILL.md (hunk 4 above)
Comment (HollowMan6) [reply skills/nemo-rl-e2e-testing/SKILL.md]: I narrowed this from general “Megatron generation/inference” coverage to an optional Megatron generation backend check, only when the NeMo-RL checkout still supports policy.generation.backend=megatron and the Bridge change explicitly targets that path. It is no longer required for normal provider compatibility.

File: skills/verl-e2e-testing/SKILL.md (hunk 5 above)
Comment (HollowMan6) [reply skills/verl-e2e-testing/SKILL.md]: I changed this to “Optional Megatron-FSDP” and clarified that FSDP is not part of the default verl validation today. It should only be run when explicitly asks for it or the change directly touches FSDP wrapping, sharding, checkpoint format, or distributed optimizer behavior.

**Review by yaoyu-33 — APPROVED** [review]
